transformer_engine/pytorch/fp8.py

```diff
@@ -477,6 +483,41 @@ def restore_fp8_meta_tensors(fp8_meta: Dict[str, Any]) -> None:
         fp8_meta["scaling_fwd"].scale_inv = fp8_meta["updated_scale_inv_fwd"]
 
 
+@contextmanager
+def fp8_init(enabled: bool = False) -> None:
+    """
+    Context manager for FP8 initialization of parameters.
+
+    .. code-block:: python
+
+        with fp8_init(enabled=True):
+            model = transformer_engine.pytorch.Linear(768, 768)
+
+    Parameters
+    ----------
+    enabled: bool, default = `False`
+                  when enabled, Transformer Engine modules created inside this `fp8_autocast`
+                  region will hold only FP8 copies of its parameters, as opposed to the default
+                  behavior where both higher precision and FP8 copies are present. Setting this
+                  option to `True` may result in lower memory consumption and is especially
+                  useful for scenarios like:
+
+                      * full model training using optimizer with master weights, where the high
+                      precision copies of weights are already present in the optimizer
+                      * inference, where only the FP8 copies of the parameters are used
+                      * LoRA-like fine-tuning, where the main parameters of the model do not
+                      change
+
+                  This functionality is *EXPERIMENTAL*.
+    """
+    try:
+        _fp8_parameters = FP8GlobalStateManager.FP8_PARAMETERS
+        FP8GlobalStateManager.FP8_PARAMETERS = enabled
+        yield
+    finally:
+        FP8GlobalStateManager.FP8_PARAMETERS = _fp8_parameters # pylint: disable=used-before-assignment
```

Comment:
This could potentially revert back to "non-fp8" state upon CM exit. 
So, doing something like:
```
with fp8_init(enabled=True):
            model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

with fp8_autocast(enabled=True):
            output = model(input) # model doesn't run in true FP8 <-----
```
because, let's say, `Linear` would call `self.primary_weights_in_fp8 = FP8GlobalStateManager.is_fp8_parameters()` which might return false once we're out of the `fp8_init` context manager? 

Unless the default value of `FP8GlobalStateManager.FP8_PARAMETERS` is `True` which wouldn't make sense either

Reply:
Or is the intended use something like:
```
with fp8_init(enabled=True):
            model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

            with fp8_autocast(enabled=True):
                            output = model(input) # model also runs in true FP8 
```
Which is fine too, but a bit inflexible? Also, will need to document this usage.

Reply:
Not sure I understand. `self.primary_weights_in_fp8` is set during constructor which happens inside `fp8_init`. Afterwards it is used irrespective of the FP8_PARAMETERS value of the context manager...

Reply:
That's right. Since we are only using the `FP8GlobalStateManager.has_fp8_parameters()` query during initialization, the value of `self.primary_weights_in_fp8` can never change during the forward execution. The only way to toggle fp8 behavior would be using the `fp8_autocast` API, which is the intended method of use.

Reply:
@sudhakarsingh27 Your intended snippet would work but the canonical (and recommended) way of going about it would be:
```
with fp8_init(enabled=True):
        model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

with fp8_autocast(enabled=True):
        output = model(input) # model also runs in true FP8
```

Reply:
We have guards to ensure that when model is initialized in the `fp8_init` context, the forward execution happens with `fp8_autocast(enabled=True)`.

Reply:
Ah you're right! I completely misunderstood `fp8_init` usage